vortex-array/src/canonical.rs

```diff
@@ -0,0 +1,471 @@
+                nulls,
+            )),
+            PType::I64 => Arc::new(LargeBinaryArray::new(
+                as_offset_buffer::<i64>(offsets),
+                data.into(),
+                nulls,
+            )),
+            _ => panic!("Invalid offsets type"),
+        },
+        DType::Utf8(_) => match offsets.ptype() {
+            PType::I32 => Arc::new(StringArray::new(
+                as_offset_buffer::<i32>(offsets),
+                data.into(),
+                nulls,
+            )),
+            PType::I64 => Arc::new(LargeStringArray::new(
+                as_offset_buffer::<i64>(offsets),
+                data.into(),
+                nulls,
+            )),
+            _ => panic!("Invalid offsets type"),
+        },
+        _ => panic!(
+            "expected utf8 or binary instead of {}",
+            varbin_array.dtype()
+        ),
+    }
+}
+
+fn local_date_time_to_arrow(local_date_time_array: LocalDateTimeArray) -> ArrayRef {
+    // A LocalDateTime maps to an Arrow Timestamp array with no timezone.
+    let timestamps = try_cast(&local_date_time_array.timestamps(), PType::I64.into())
+        .expect("timestamps must cast to i64")
+        .into_canonical()
+        .and_then(Canonical::into_primitive)
+        .expect("must be i64 array");
+    let validity = timestamps
+        .logical_validity()
+        .to_null_buffer()
+        .expect("null buffer");
+    let timestamps_len = timestamps.len();
+    let buffer = ScalarBuffer::<i64>::new(timestamps.into_buffer().into(), 0, timestamps_len);
+
+    match local_date_time_array.time_unit() {
+        TimeUnit::Ns => Arc::new(TimestampNanosecondArray::new(buffer, validity)),
+        TimeUnit::Us => Arc::new(TimestampMicrosecondArray::new(buffer, validity)),
+        TimeUnit::Ms => Arc::new(TimestampMillisecondArray::new(buffer, validity)),
+        TimeUnit::S => Arc::new(TimestampSecondArray::new(buffer, validity)),
+    }
+}
+
+/// Support trait for transmuting an array into its [vortex_dtype::DType]'s canonical encoding.
+///
+/// This conversion ensures that the array's encoding matches one of the builtin canonical
+/// encodings, each of which has a corresponding [Canonical] variant.
+///
+/// # Invariants
+///
+/// The DType of the array will be unchanged by canonicalization.
+pub trait IntoCanonical {
+    fn into_canonical(self) -> VortexResult<Canonical>;
+}
+
+/// Trait for types that can be converted from an owned type into an owned array variant.
+///
+/// # Canonicalization
+///
+/// This trait has a blanket implementation for all types implementing [IntoCanonical].
+pub trait IntoArrayVariant {
+    fn into_null(self) -> VortexResult<NullArray>;
+
+    fn into_bool(self) -> VortexResult<BoolArray>;
+
+    fn into_primitive(self) -> VortexResult<PrimitiveArray>;
+
+    fn into_struct(self) -> VortexResult<StructArray>;
+
+    fn into_varbin(self) -> VortexResult<VarBinArray>;
+
+    fn into_extension(self) -> VortexResult<ExtensionArray>;
+}
```

Comment:
I added this as a new trait with blanket impl over `T: IntoCanonical`.

I wanted to do it as methods on `IntoCanonical` with default impls, but that actually fails because `self` is unsized, and you can't make it sized b/c then `ArrayTrait` is no longer object-safe. Now that I've done this, I actually think it's better than default methods